subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c

```diff
@@ -251,6 +254,82 @@ int nrf_cloud_decode_requested_state(const struct nrf_cloud_data *input,
 	return 0;
 }
 
+int nrf_cloud_encode_config_response(struct nrf_cloud_data const *const input,
+				     struct nrf_cloud_data *const output,
+				     bool *const has_config)
+{
+	__ASSERT_NO_MSG(output != NULL);
+	char *buffer;
+	cJSON *root_obj = cJSON_CreateObject();
+	cJSON *desired_obj = cJSON_CreateObject();
+	cJSON *null_obj = cJSON_CreateNull();
+	cJSON *input_obj = input ? cJSON_Parse(input->ptr) : NULL;
+	cJSON *reported_obj = input_obj ? cJSON_CreateObject() : NULL;
+	cJSON *state_obj = NULL;
+	cJSON *config_obj = NULL;
+
+	if ((root_obj == NULL) || (desired_obj == NULL) || (null_obj == NULL) ||
+	    (input_obj && !reported_obj)) {
+		/* A NULL input_obj is not a memory error */
+		cJSON_Delete(root_obj);
+		cJSON_Delete(desired_obj);
+		cJSON_Delete(null_obj);
+		cJSON_Delete(input_obj);
+		cJSON_Delete(reported_obj);
+		return -ENOMEM;
+	}
+
+	/* A delta update will have state */
+	state_obj = cJSON_DetachItemFromObject(input_obj, "state");
+	config_obj = cJSON_DetachItemFromObject(
+		state_obj ? state_obj : input_obj, "config");
+	cJSON_Delete(input_obj);
+
+	if (has_config) {
+		*has_config = (config_obj != NULL);
+	}
+
+	/* A state and config indicate this is a delta */
+	if (state_obj && config_obj) {
+		/* Add delta config to reported */
+		(void)json_add_obj(reported_obj, "config", config_obj);
+		(void)json_add_obj(root_obj, "reported", reported_obj);
+
+		/* Add a null config to desired */
+		(void)json_add_obj(desired_obj, "config", null_obj);
+		(void)json_add_obj(root_obj, "desired", desired_obj);
+
+		/* Cleanup received state obj and re-use for the response */
+		cJSON_Delete(state_obj);
+		state_obj = cJSON_CreateObject();
+		(void)json_add_obj(state_obj, "state", root_obj);
+		buffer = cJSON_PrintUnformatted(state_obj);
+		cJSON_Delete(state_obj);
+
+		if (buffer == NULL) {
+			return -ENOMEM;
```

Comment:
Don't we need to cleanup the root object and friends as well?

Reply:
root and friends get added to `state_obj`.  deleting `state_obj` will take care of root and friends.